core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -173,10 +178,13 @@ public void handleResolvedAddressGroups(
       // Shutdown subchannels for removed addresses.
       for (EquivalentAddressGroup addressGroup : removedAddrs) {
         Subchannel subchannel = subchannels.remove(addressGroup);
-        subchannel.shutdown();
+        update = update || isReady(subchannel); // no need to update if channel was already excluded
```

Comment:
Isn't this missing for the `addedAddrs` ?  i.e. why only update on removing addresses?

Reply:
The main intention is to avoid calling `updateBalancingState()` in cases that won't result in any change to the Picker state, particularly when it already has active endpoints.

Any new `addedAddrs` are added in `IDLE` state and so will have no effect on the Picker state if already has at least one `READY` subchannel (and if it doesn't then `ready` will be false and so we update anyhow).